drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1195 @@
+		goto out_error;
+	}
+	fprintf(stderr, "mode for connector %u is %ux%u\n",
+	        conn->connector_id, out->bufs[0].width, out->bufs[0].height);
+
+	/* find a crtc for this connector */
+	ret = modeset_find_crtc(fd, res, conn, out);
+	if (ret) {
+		fprintf(stderr, "no valid crtc for connector %u\n",
+			conn->connector_id);
+		goto out_blob;
+	}
+
+	/* with a connector and crtc, find a primary plane */
+	ret = modeset_find_plane(fd, out);
+	if (ret) {
+		fprintf(stderr, "no valid plane for crtc %u\n", out->crtc_id);
+		goto out_blob;
+	}
+
+	/* gather properties of our connector, CRTC and planes */
+	ret = modeset_setup_objects(fd, out);
+	if (ret) {
+		fprintf(stderr, "cannot get plane properties\n");
+		goto out_blob;
+	}
+
+	/* setup front/back framebuffers for this CRTC */
+	ret = modeset_setup_framebuffers(fd, conn, out);
+	if (ret) {
+		fprintf(stderr, "cannot create framebuffers for connector %u\n",
+			conn->connector_id);
+		goto out_obj;
+	}
+
+	return out;
+
+out_obj:
+	modeset_destroy_objects(fd, out);
+out_blob:
+	drmModeDestroyPropertyBlob(fd, out->mode_blob_id);
+out_error:
+	free(out);
+	return NULL;
+}
+
+/*
+ * modeset_prepare() changes a little bit. Now we use the new function
+ * modeset_output_create() to allocate memory and setup the output.
+ */
+
+static int modeset_prepare(int fd)
+{
+	drmModeRes *res;
+	drmModeConnector *conn;
+	unsigned int i;
+	struct modeset_output *out;
+
+	/* retrieve resources */
+	res = drmModeGetResources(fd);
+	if (!res) {
+		fprintf(stderr, "cannot retrieve DRM resources (%d): %m\n",
+			errno);
+		return -errno;
+	}
+
+	/* iterate all connectors */
+	for (i = 0; i < res->count_connectors; ++i) {
+		/* get information for each connector */
+		conn = drmModeGetConnector(fd, res->connectors[i]);
+		if (!conn) {
+			fprintf(stderr, "cannot retrieve DRM connector %u:%u (%d): %m\n",
+				i, res->connectors[i], errno);
+			continue;
+		}
+
+		/* create an output structure and free connector data */
+		out = modeset_output_create(fd, res, conn);
+		drmModeFreeConnector(conn);
+		if (!out)
+			continue;
+
+		/* link output into global list */
+		out->next = output_list;
+		output_list = out;
+	}
+	if (!output_list) {
+		fprintf(stderr, "couldn't create any outputs\n");
+		return -1;
+	}
+
+	/* free resources again */
+	drmModeFreeResources(res);
+	return 0;
+}
+
+/*
+ * modeset_atomic_prepare_commit() is new. Here we set the values of properties
+ * (of our connector, CRTC and plane objects) that we want to change in the
+ * atomic commit. These changes are temporarily stored in drmModeAtomicReq *req
+ * until the commit actually happens.
+ */
+
+static int modeset_atomic_prepare_commit(int fd, struct modeset_output *out,
+					 drmModeAtomicReq *req)
+{
+	struct drm_object *plane = &out->plane;
+	struct modeset_buf *buf = &out->bufs[out->front_buf ^ 1];
+
+	/* set id of the CRTC id that the connector is using */
+	if (set_drm_object_property(req, &out->connector, out->conn_id,
```

Comment:
If object ID was a field in `struct drm_object`, all these set property calls would be shorter. Another reason why the field would be a good idea is that the properties are specific to the particular object (ID) and should not be attempted to be used for any other object, even of the same type.

Reply:
Done.